main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -37,13 +45,233 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
     */
   def run(root: Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = {
 
-    val b = System.nanoTime()
+    val startTime = System.nanoTime()
+
+    val stratified = stratify(root.strata, root.tables.keys.toList)
+    val duration = System.nanoTime() - startTime
+
+    for {
+      stratified <- stratified
+    } yield root.copy(strata = stratified).copy(time = root.time.copy(stratifier = duration))
+  }
+
+  /**
+    * Stratify the graph
+    */
+  def stratify(stratum: List[Stratum], syms: List[Symbol.TableSym]): Validation[List[Stratum], StratificationError] = {
+    // Implementing as described in Database and Knowledge - Base Systems Volume 1
+    // Ullman, Algorithm 3.5 p 133
+
+    /// Start by creating a mapping from predicates to stratum
+    var strataMap = syms.map(x => (x, 1)).toMap
```

Comment:
Perhaps rename the map to `stratumOf` since calling it `Map` does not really offer much information.